:lock_with_ink_pen: Implement Dynamic Winternitz Commitment Scheme (#23)
distributed-lab/nero: docs/paper/nero.tex

        rec_size += int(i * w)
    
    total = pk_size + ver_size + rec_size + sig_size
    
    print(f"{d} & {sig_size} & {pk_size} & {ver_size} & {rec_size} & {total} \\\\")
\end{verbatim}
\fi
\begin{table}[H]
  \centering
  \begin{tabular}{cccccc}
    \toprule
    $d$ & \textbf{Signature} & \textbf{Public Key} & \textbf{Verification Script} & \textbf{Recovery Script} & \textbf{Total} \\
    \midrule
    3 & 420 & 400 & 120 & 240 & 1180 \\
    7 & 294 & 280 & 196 & 165 & 935 \\
    15 & 210 & 200 & 300 & 112 & 822 \\
    31 & 189 & 180 & 558 & 105 & 1032 \\
    63 & 168 & 160 & 1008 & 90 & 1426 \\
    127 & 147 & 140 & 1778 & 70 & 2135 \\
    255 & 126 & 120 & 3060 & 48 & 3354 \\
    \bottomrule
  \end{tabular}
  \caption{Different script sizes depending on the $d$ value per each
    32-bit message. Note that, ``Signature'' column includes the
    encoding of 32-bit message in it.}\label{tab:winternitz-script-size}
\end{table}

\subsubsection{Compact Winternitz commitment
  scheme in Bitcoin}\label{sec:compact-winternitz-in-bitcoin}

As we mentioned in prevoius section, the most contribution to script
size comes from public key, signature, verification and recovery, but

> do we really need to do that for all of the limbs?

Reviewer: ```suggestion
do we actually need to sign each integer limb?
```

Let's check the checksum of winternitz singature described
in~\cite{applied-crypto}:

\begin{equation}
  c \gets d n_0 - (e_1 + \ldots + e_{n_0})

> \end{equation}\label{eq:winternitz-checksum}

Reviewer: If an equation is small, I prefer writing it inline (this is the case). Also, instead of `\gets`, it is OK to use the `=` sign here (typically, `\gets` is used for describing algorithms where you set/change some variables in the process/in the protocol).

Author: I just inherited it from the book we are referring to, so not sure we need to make any changes here

> Assuming that: $d=15$ and $e_{7} = 0$ --- is the last limb of the
> $\mathsf{Enc}(m)$, from~\eqref{eq:winternitz-checksum} the value of
> $e_{7}$ is obviously negliable in the sum of the checksum, so, without
> any security concerns, can be skipped. But to keep the right value

Reviewer: ```suggestion
Suppose, at some point during the protocol execution with $d=15$, the last limb of the message encoding $\mathsf{Enc}(m)$ turns out to be $e_{7} = 0$. According to \eqref{eq:winternitz-checksum}, the value of
$e_{7}$ does not contribute to the sum of the checksum, so, without any security concerns, can be omitted. But to keep the right value
```

> after a recover, we can't remove all ``zeroed'' limbs. That's why

Reviewer: ```suggestion
after a recover, we can't remove all zero limbs. That's why
```

> instead, we propose \textit{skipping} only the most significant limbs

Reviewer: ```suggestion
instead, we propose \textit{skipping} only the most significant \textit{zero} limbs
```

of a stack element.

> This makes the recovery script size equal to zero in best case
> scenarios, public key's and signature's hashes number from $1 + n_1$
> in best cases to $N$ in worst, and the same applies to verification
> script size too, as less checks required for smaller public key.

Reviewer: ```suggestion
This makes the recovery script size equal to zero in the best-case scenario. The public key and signature hash digest number ranges from $1 + n_1$ in the best case to $N$ in the worst. The same applies to the total verification script size, as fewer checks are required for smaller public keys.
```

Reviewer: Also, make sure to update the Checklist after considering the suggestions
